cardano-crypto-class/src/Cardano/Crypto/DSIGN/SchnorrSecp256k1.hs

```diff
@@ -0,0 +1,296 @@
+  plusForeignPtr, 
+  castForeignPtr
+  )
+import Control.Monad (when)
+import System.IO.Unsafe (unsafeDupablePerformIO, unsafePerformIO)
+import Cardano.Binary (FromCBOR (fromCBOR), ToCBOR (toCBOR, encodedSizeExpr))
+import Foreign.Ptr (Ptr, castPtr, nullPtr)
+import Foreign.C.Types (CUChar)
+import Foreign.Marshal.Alloc (allocaBytes)
+import Cardano.Crypto.Schnorr (
+  SECP256k1XOnlyPubKey, 
+  secpKeyPairCreate, 
+  SECP256k1Context, 
+  secpKeyPairXOnlyPub,
+  SECP256k1SecKey, 
+  secpSchnorrSigVerify, 
+  secpContextSignVerify,
+  SECP256k1SchnorrSig, 
+  secpSchnorrSigSignCustom, 
+  secpContextCreate
+  )
+import NoThunks.Class (NoThunks, OnlyCheckWhnfNamed (OnlyCheckWhnfNamed))
+import Cardano.Crypto.DSIGN.Class (
+  DSIGNAlgorithm (VerKeyDSIGN, 
+                  SignKeyDSIGN, 
+                  SigDSIGN,
+                  SeedSizeDSIGN, 
+                  SizeSigDSIGN, 
+                  SizeSignKeyDSIGN, 
+                  SizeVerKeyDSIGN, 
+                  algorithmNameDSIGN,
+                  deriveVerKeyDSIGN, 
+                  signDSIGN, 
+                  verifyDSIGN, 
+                  genKeyDSIGN, 
+                  rawSerialiseSigDSIGN,
+                  Signable, 
+                  rawSerialiseVerKeyDSIGN, 
+                  rawSerialiseSignKeyDSIGN, 
+                  rawDeserialiseVerKeyDSIGN,
+                  rawDeserialiseSignKeyDSIGN, 
+                  rawDeserialiseSigDSIGN), 
+  encodeVerKeyDSIGN, 
+  encodedVerKeyDSIGNSizeExpr, 
+  decodeVerKeyDSIGN, 
+  encodeSignKeyDSIGN, 
+  encodedSignKeyDESIGNSizeExpr, 
+  decodeSignKeyDSIGN, 
+  encodeSigDSIGN, 
+  encodedSigDSIGNSizeExpr, 
+  decodeSigDSIGN
+  )
+import Cardano.Crypto.Util (SignableRepresentation (getSignableRepresentation))
+
+data SchnorrSecp256k1DSIGN
+
+instance DSIGNAlgorithm SchnorrSecp256k1DSIGN where
+  type SeedSizeDSIGN SchnorrSecp256k1DSIGN = 32
+  type SizeSigDSIGN SchnorrSecp256k1DSIGN = 64
+  type SizeSignKeyDSIGN SchnorrSecp256k1DSIGN = 32
+  type SizeVerKeyDSIGN SchnorrSecp256k1DSIGN = 64
+  type Signable SchnorrSecp256k1DSIGN = SignableRepresentation
+  newtype VerKeyDSIGN SchnorrSecp256k1DSIGN = 
+    VerKeySchnorr256k1 (ForeignPtr SECP256k1XOnlyPubKey)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "VerKeySchnorr256k1" (ForeignPtr SECP256k1XOnlyPubKey)
+        )
+  newtype SignKeyDSIGN SchnorrSecp256k1DSIGN = 
+    SignKeySchnorr256k1 (ForeignPtr SECP256k1SecKey)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "SignKeySchnorr256k1" (ForeignPtr SECP256k1SecKey)
+        )
+  newtype SigDSIGN SchnorrSecp256k1DSIGN = 
+    SigSchnorr256k1 (ForeignPtr SECP256k1SchnorrSig)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "SigSchnorr256k1" (ForeignPtr SECP256k1SchnorrSig)
+        )
+  algorithmNameDSIGN _ = "schnorr-secp256k1"
+  {-# NOINLINE deriveVerKeyDSIGN #-}
+  deriveVerKeyDSIGN (SignKeySchnorr256k1 fp) = 
+    unsafeDupablePerformIO . withForeignPtr fp $ \skp -> do
```

Comment:
Do we want to isolate this stuff in a module that provides a nicer Haskell interface, more like `secp256k1-haskell` provides?

Reply:
I don't feel this would help much. When it comes to FFI stuff, my policy is to bind as closely as possible to the C API, then wrap it directly. Adding another level of indirection in between doesn't strike me as useful, and has actually backfired quite badly in the past.